### PR TITLE
Auth: Fix orgrole picker disabled if isSynced user

### PR DESCRIPTION
**Pull request by eleijonmarck**
**What is this feature?**
- disables the orgrole picker if the user is synced from external provider
- disables the basicRole picker for role picker if the users is synced
- adds header to column with "Synced from"
- backend checks for externalUser info
- add tests on backend for IsExternallySynced
- implemented feature toggle for this breaking change, onlyExternalOrgRoleSync

TODO:
- [x] implement featuretoggle for this
- [x] Currently resetting the particular User's entry/entries in user_auth table is not supported. Thus, I think the best we can do on the short term to implement what @Jguer suggested: when a provider is disabled should be to always return not ExternallySync.
- [x] update doc for Grafana 10 we remove the feature toggle
- [x] reach out to support
- [x] runbook available in case supports needs help - https://github.com/grafana/grafana-authnz-team/wiki/Orgrole-picker-disabled-in-users-page

Fixes #
https://github.com/grafana/grafana/issues/63835

**Special notes for your reviewer**:

Goes together w. enterprise - https://github.com/grafana/grafana-enterprise/pull/4774

oss test
image

enterprise test
image

### CODE REVIEW

**Comment by mgyongyosi** [comment]
Could you please ensure that the role cannot be changed through the API by adding a check to the backend as well? AFAIK there isn't any check implemented on the backend, but I'm not sure.

**Comment by eleijonmarck** [comment]
> Could you please ensure that the role cannot be changed through the API by adding a check to the backend as well? AFAIK there isn't any check implemented on the backend, but I'm not sure.

that's a valid point

I'l make another PR for that check if that's alright?

**Comment by mgyongyosi** [comment]
> > Could you please ensure that the role cannot be changed through the API by adding a check to the backend as well? AFAIK there isn't any check implemented on the backend, but I'm not sure.
> 
> that's a valid point
> 
> I'l make another PR for that check if that's alright?

I think it would be better to handle the backend check as part of this PR because that should be the part of the fix. And having only one PR would make it trivial to find the related changes in the future.

File: public/app/features/admin/UserOrgs.tsx
Comment (linoman) [inline public/app/features/admin/UserOrgs.tsx:421]: 🎉

**Review by linoman — APPROVED** [review]

File: pkg/services/org/orgimpl/store.go
Comment (Jguer) [inline pkg/services/org/orgimpl/store.go:504]: careful not to merge debug prints

Comment (eleijonmarck) [reply pkg/services/org/orgimpl/store.go:504]: Yea I will not Merge until I get the backend ready. It's work in progress. 

Il mark it as Draft

**Comment by Jguer** [comment]
Non-blocking question we should have an answer to: If I stop using an auth provider to login and return to basic auth, what should I do to be able to regain control over the user's attributes?

File: pkg/api/org_users.go
Comment (Jguer) [inline pkg/api/org_users.go:398]: Missed it last review but shouldn't `IsExternallySynced` also confirm the provider is enabled in the cfg, so as to not block users from changing from provider -> no provider.

Behavior of when a provider is disabled should be to always return not ExternallySync

Comment (Jguer) [reply pkg/api/org_users.go:398]: Also occasion to fix `AuthProxtLabel` typo in `pkg/services/login/authinfo.go` 

EDIT: nit nit:`Github` -> `GitHub`

**Review by Jguer — CHANGES_REQUESTED** [review]

File: pkg/api/org_users.go
Comment (eleijonmarck) [reply pkg/api/org_users.go:398]: the Github change will include all the settings as well. all of the settings are Github instead of GitHub. I'l see if I want to make that change here

Comment (Jguer) [reply pkg/api/org_users.go:398]: Yeah, do not bother with the GitHub if it's much work

Comment (eleijonmarck) [reply pkg/api/org_users.go:398]: > Behavior of when a provider is disabled should be to always return not ExternallySync

this is true. I think we do not have all settings for that. Is it possible to add that in a separate PR?

**Comment by eleijonmarck** [comment]
> Non-blocking question we should have an answer to: If I stop using an auth provider to login and return to basic auth, what should I do to be able to regain control over the user's attributes?

@Jguer I think the answer to this, is that we reset the AuthProvider to use the "latest" that you logged in with. Not sure how we deal with that to reset the AuthProvider.

**Comment by eleijonmarck** [comment]
@Jguer changed milestone to 9.5 as this is breaking change and we can make some more copy and documentation around it

**Comment by eleijonmarck** [comment]
added breaking change label as part of this PR

File: docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md:93]:
```suggestion
| `onlyExternalOrgRoleSync`          |  Prohibits a user from changing organization roles synced with external auth providers                                                                                      |
```

**Review by chri2547 — DISMISSED** [review]
Added suggestion. Thank you!

File: docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md
Comment (eleijonmarck) [reply docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md:93]: this file is generated automatically, so i will change it in the registry of the featuretoggle

File: pkg/api/org_users.go
Comment (eleijonmarck) [reply pkg/api/org_users.go]: made it part of this PR - [85635cc](https://github.com/grafana/grafana/pull/64033/commits/85635cc1433a2fb8d5472958dff38eb26b106efb)

File: pkg/services/featuremgmt/registry.go
Comment (eleijonmarck) [inline pkg/services/featuremgmt/registry.go:417]: naming things! anyone feel like they have a good suggestion?

We want to make the name as a opt-in

**Review by chri2547 — COMMENTED** [review]
Docs signoff complete, LGTM!

File: pkg/api/org_users.go
Comment (Jguer) [inline pkg/api/org_users.go:400]: `IsExternallySynced` is using Auth Label to check settings while here it's being provided AuthModule.

Backend check will always return false here.

Change `IsExternallySynced` to always use the AuthModule and not the frontend only Auth Label

File: pkg/services/login/authinfo.go
Comment (Jguer) [inline pkg/services/login/authinfo.go:135]: This setting is never read in `readGenericOAuthSettings` 

As such Generic OAuth label will always allow changes 



Extra: It should also be used in `pkg/login/social/social.go` when configuring the Generic oauth to prevent changing config key names from presenting an inaccurate system state. But this can be done later when social is refactored

**Review by Jguer — DISMISSED** [review]

File: pkg/api/org_users_test.go
Comment (eleijonmarck) [reply pkg/api/org_users_test.go:261]: 🤦 fixing this !

File: pkg/services/login/authinfo.go
Comment (eleijonmarck) [reply pkg/services/login/authinfo.go:135]: this one is interesting, so they can configure and change the settings. we should make sure to capture that. I guess one thing we need to do as well is to take the env variable in case they spin it up with docker.

Il add that

File: public/app/features/users/UsersTable.tsx
Comment (IevaVasiljeva) [inline public/app/features/users/UsersTable.tsx:53]: nit: this looks a bit odd if no external auth providers are used.

**Review by IevaVasiljeva — CHANGES_REQUESTED** [review]
I've tested this with LDAP and SAML and basic auth, and covered the following scenarios:
1) no external auth provider used;
2) one external auth provider, `skipOrgRoleSync` enabled;
3) one external auth provider, `skipOrgRoleSync` disabled;
3) two external auth providers, `skipOrgRoleSync` enabled for one and disabled for the other.

All works fine apart from the basic auth scenario where there's no entries in the `user_auth` table. Once that is fixed, I think we're good to merge :)

File: pkg/services/org/model.go
Comment (eleijonmarck) [reply pkg/services/org/model.go:20]: i wanted to use this in the `org_users` response.Error, but it is better to follow the conventions there. nice catch 👍

File: public/app/features/users/UsersTable.tsx
Comment (eleijonmarck) [reply public/app/features/users/UsersTable.tsx:53]: il see if i could change the wording. last take would be if we have externally synced users and show it

Comment (eleijonmarck) [reply public/app/features/users/UsersTable.tsx:53]: changed it to "Origin" instead of "synced"

File: pkg/services/org/model.go
Comment (eleijonmarck) [reply pkg/services/org/model.go:20]: used in the error handling now for the API call

File: pkg/api/org_users.go
Comment (eleijonmarck) [reply pkg/api/org_users.go:396]: image

basic auth users should be able to change the role now. It is only applicable for external users. The code looks a bit ugly, will see if I could refactor before a new review

Comment (eleijonmarck) [reply pkg/api/org_users.go:396]: potential refactor would be to allow the IsExternallySynced to take in a userID as a parameter instead. This would clean up this code and one could be able to use it with only the ID of the user without needing to know If there is a AuthModule attached.

leaving this as a potential refactor for later

File: pkg/api/org_users_test.go
Comment (eleijonmarck) [inline pkg/api/org_users_test.go:256]: added tests for basic Auth as well

File: pkg/api/user.go
Comment (eleijonmarck) [inline pkg/api/user.go:70]: using AuthModule instead of Labels

File: public/app/features/admin/UserListAdminPage.tsx
Comment (eleijonmarck) [inline public/app/features/admin/UserListAdminPage.tsx:153]: changed to Origin instead, wdyt?

**Review by IevaVasiljeva — APPROVED** [review]
I've just tested it again, and everything works great!

Please lower the level of the log before merging (and maybe fix the way the error is used). But apart from that this looks good 👌

**Comment by jdbaldry** [comment]
@eleijonmarck Should this be backported to `v9.5.x`. It has the `9.5.0` milestone.

**Pull request merged**
